week.go

```diff
@@ -39,26 +40,96 @@ func (w *Week) Week() int {
 // Next calculates and returns the next week. If the next week is invalid (year > 9999) the function
 // returns an error.
 func (w *Week) Next() (Week, error) {
-	var year, week int
-
-	if w.week >= weeksInYear(w.year) {
-		year, week = w.year+1, 1
-	} else {
-		year, week = w.year, w.week+1
-	}
-
-	return New(year, week)
+	return w.Add(1)
 }
 
 // Previous calculates and returns the previous week. If the previous week is invalid (year < 0) the
 // function returns an error.
 func (w *Week) Previous() (Week, error) {
-	var year, week int
+	return w.Add(-1)
+}
+
+// Add calculates and returns a week that is the given positive distance (number of weeks) from the current week
+func (w *Week) Add(weeks int) (Week, error) {
+	return w.add(weeks)
+}
+
+// Sub calculates the positive difference between w and u (u-w) in number of weeks
+func (w *Week) Sub(u *Week) int {
+	var (
+		diff        = 0
+		yearDiff    = u.year - w.year
+		direction   = 1
+		smallerWeek = w
+		biggerWeek  = u
+	)
+
+	if yearDiff != 0 {
+		direction = int(math.Sqrt(float64(yearDiff*yearDiff))) / yearDiff
+	}
+
+	if direction == -1 {
+		smallerWeek = u
+		biggerWeek = w
+	}
+
+	var (
+		yearA = smallerWeek.year
+		yearB = biggerWeek.year
+		weekA = smallerWeek.week
+		weekB = biggerWeek.week
+	)
+
+	for {
+		if yearA > yearB {
+			panic("infinite loop guard: yearA should never be more than yearB")
+		}
+
+		if yearA != yearB {
+			diff += weeksInYear(yearA)
+			yearA++
+			continue
+		}
+
+		diff += weekB - weekA
+		break
+	}
+
+	return diff * direction
+}
+
+func (w *Week) add(weeksToAdd int) (Week, error) {
+	var sign = 1
+
+	if weeksToAdd < 0 {
+		sign = -1
+	}
+
+	var year = w.year
+	var week = w.week + weeksToAdd
+	var maxWeeks = weeksInYear(w.year)
+
+	for {
+		if week > maxWeeks || week < 0 {
```

Comment:
In Go it is very common to do the end / error conditions first and keep the happy path on the main indention level.